### ISSUE TITLE
Offline learning in Presage

### ISSUE THREAD

**Issue by kmcnaught** [OptiKey/OptiKey#795]
This is more a question, though it might lead to a feature request. 

I've been re-investigating the Presage prediction engine recently, and haven't worked out whether there is currently any way to load up any pre-existing text for presage to learn from? The presage website states that it has:
> adaptive online and offline learning

but no information about whether the offline learning is exposed to an end user or just an API-consumer. I can hack something into `PresageSuggestions.cs` to read a text file and call `prsg.learn(my_text)` as a one-time thing, and this gives me the result I'm after, but this isn't something easily available to users.

1 - Am I missing something? Has anyone found an easier way to pre-train presage?

2 - If there isn't already, we might consider adding this functionality into Optikey, for example via the management console.

**Comment by JuliusSweetland** [comment]
@kmcnaught This is from memory, but I think when you want to create a Presage database for a new language there's a process of training it on example text(s). Perhaps this is what you want?

**Comment by kmcnaught** [comment]
Ah okay, following your wiki notes to https://github.com/Manouchehri/presage/blob/master/FAQ it looks like you can use the `text2ngram` program to train a model on your own text to begin with. That would have solved my initial problem, but I've been thinking about how this could be even more helpful. What I'm imagining is that you might want to regularly (say, once a day or once a week) load up new (additional) training data into your existing presage model so that you can work in a new context quickly. For example, if I need to write a report and I have some old template reports with similar material in, I may want to pre-emptively load up the vocabulary into the recency model. Or I want to discuss this week's lectures with a fellow student, so I use the lecture transcript to get all the vocabulary. 

This is actually trivial to implement if we were happy to have a static instance of the presage engine and a new FunctionKey to _"add clipboard text to presage"_. Otherwise, there are too many layers of separation between the mainviewmodel and the actual presage instance.

**Comment by kmcnaught** [comment]
I've been doing this in another fork, using a static instance of the presage engine. What still makes me uncomfortable is that I'm not sure what happens when you ask Presage repeatedly to learn from the same text. It would be nice to be able to manually seed the context, but the underlying mechanism is unclear to me and I'm not sure if we'll end up with a poorly behaving model in Presage
